src/artifact/decl.rs

```diff
@@ -29,15 +28,145 @@ impl ImportKind {
     }
 }
 
+#[derive(Debug, Copy, Clone, PartialEq, Eq, PartialOrd, Ord)]
+/// Linker binding scope of a definition
+pub enum Scope {
+    /// Available to all components
+    Global,
+    /// Available only inside the defining component
+    Local,
+    /// Available to all modules, but only selected if a Global
+    /// definition is not found
+    Weak,
+}
+
+macro_rules! scope_methods {
+    () => {
+    /// Set scope to global
+    pub fn global(self) -> Self {
+        self.with_scope(Scope::Global)
+    }
+    /// Set scope to local
+    pub fn local(self) -> Self {
+        self.with_scope(Scope::Local)
+    }
+    /// Set scope to weak
+    pub fn weak(self) -> Self {
+        self.with_scope(Scope::Weak)
+    }
+    /// Builder for scope
+    pub fn with_scope(mut self, scope: Scope) -> Self {
+        self.scope = scope;
+        self
+    }
+    /// Gst scope
+    pub fn get_scope(&self) -> Scope {
+        self.scope
+    }
+    /// Set scope
+    pub fn set_scope(&mut self, scope: Scope) {
+        self.scope = scope;
+    }
+    /// Check if scope is `Scope::Global`
```

Comment:
A possible confusion here is whether weak symbols count as "global", since they act like they're globals in most respects. I don't have a strong opinion about what `is_global` should do, but it'd be good to explicitly mention what this means for weak symbols in the comment.